### ISSUE TITLE
Create plan(s) for stopflow experiment

### ISSUE THREAD

**Issue by callumforrester** [DiamondLightSource/i22-bluesky#13]
Write a new exemplar plan for running a stopflow experiment, to be refined for the June experiment.

## Experiment Specification

The experiment involves a liquid sample intersecting and flowing laterally to the beam. The beam passes through the flowing liquid and scatters onto a detector. The flow is controlled by a proprietary rig that allows the user to request a "stop flow". At the moment the flow stops, the rig sends out a pulse to the panda, which can then trigger the detector.

![stop-flow-experiment(2)](https://github.com/DiamondLightSource/i22-bluesky/assets/29771545/bf781a89-df61-4c49-9b6e-939402333fbb)

A single pulse goes into the panda and many pulses come out via a sequencer table. The table can specify X sets of Y pulses at Z hertz.

**Comment by timsnow** [comment]
It may also be that a user wishes to collect a number of frames of data before waiting for the trigger as well.

Which leaves us with the following possibilities:

- Start acquisition -> wait for trigger -> acquire m frames
- Start acquisition -> acquire n frames -> wait for trigger -> acquire m frames

**Comment by coretl** [comment]
@timsnow also what happens to the fast shutter during the wait period? Do we close it while waiting, then open it on TTL trigger, or keep it open and trigger the detector immediately on that TTL trigger?

**Comment by stan-dot** [comment]
@callumforrester we need a formal specification, I'm on it to make a state diagram. 
https://mermaid.js.org/syntax/stateDiagram.html

**Comment by coretl** [comment]
The sequence is:
- Create a SequenceTable object that is:
  - Send n triggers
  - Wait for BITA=1
  - Send m triggers
- Prepare detector for n + m frames, and panda with the SequenceTable
- Kickoff and Complete and periodic collects as usual
- After n frames have been collected, the user can press the button to stop the flow, and m frames will be collected

Considerations:
- We timeout if a frame has not been collected in 60s, is this long enough to press the stop flow button or should we turn this off?
- Should we have the fast shutter open for the whole section, or close it while waiting for user to press the button?

**Comment by coretl** [comment]
In P45 use the web GUI to set `SEQ1.BITA` to `ONE` then back to `ZERO` to simulate the stop cell, in I22 this will be connected to a `TTLIN`

**Comment by coretl** [comment]
This is how P45 is wired up to support this:
![image](https://github.com/DiamondLightSource/i22-bluesky/assets/101418278/fa11abb0-a1cd-42cf-a390-32afc187854b)

**Comment by timsnow** [comment]
> Questions:
> 
> * We timeout if a frame has not been collected in 60s, is this long enough to press the stop flow button or should we turn this off?
> * Are the n and the m frames at the same rate?
> * Should we have the fast shutter open for the whole section, or close it while waiting for user to press the button?
> * What do we want to capture on the PandA? Just TS_START and GATE_DURATION?

Answers:

* 60 seconds should be enough, however, does this apply _after_ the sequencer table has sent out the first gate? (_i.e._ would PCAP throw an error mid-measurement if it has to wait more than 60 seconds?)
* Typically yes, however, it is not impossible that _m_ and _n_ would be of different frame rates
* The fast shutter should be triggered 4 ms before the first frame (or burst of frames should the fast shutter need to be opened and closed more than once during measurements) but only once measurements have begun - not purely at the start of the scan
* For now, yes, in the fullness of time it would be useful to have the times that the fast shutter is opened and closed as well as when the detectors were triggered

**Comment by timsnow** [comment]
Questions:

> > ```
> > * 60 seconds should be enough, however, does this apply _after_ the sequencer table has sent out the first gate? (_i.e._ would PCAP throw an error mid-measurement if it has to wait more than 60 seconds?)
> > ```
> 
> Yes, if the sequencer table has started, then it waits for up to the timeout for a frame to arrive, so if you made `m` frames, then waited for 70s for the stop flow to get a trigger, then it would fail.
> 
> This is the code that applies the timeout: https://github.com/bluesky/ophyd-async/blob/9f332e389da01b3841c7d000ee7ddfa13c7b3502/src/ophyd_async/core/detector.py#L311-L322
> 
> The default timeout is actually 10s rather than 60s, we need to bump that up in this case, or think of a better way of doing this.
> 
> > ```
> > * Typically yes, however, it is not impossible that _m_ and _n_ would be of different frame rates
> > ```
> 
> For June, can we assume they are the same?
> 
> > ```
> > * The fast shutter should be triggered 4 ms before the first frame (or burst of frames should the fast shutter need to be opened and closed more than once during measurements) but only once measurements have begun - not purely at the start of the scan
> > ```
> 
> So while you are waiting for stop flow to send a trigger, should the shutter be open or closed?
> 
> > ```
> > * For now, yes, in the fullness of time it would be useful to have the times that the fast shutter is opened and closed as well as when the detectors were triggered
> > ```
> 
> I can't actually think of how to do this, but it's not needed for June so I'll leave this for now

Answers:

* An initial wait time (_i.e._) knowing that things have started is fine, this has been the default behaviour for the TFG/Pilatus setup, however, once the first trigger has been sent the wait can be almost infinite - there are definitely times where you could be waiting minutes, or tens of minutes, for the second trigger, _e.g._ sample heating stage going from 10-100 degrees celsius with an equilibration time
* As different frame times is the next use case for the pressure cell, yes we can assume that _m_ and _n_ are equal for June
* Good question, I can imagine a scenario for both, however as you're typically expecting fast kinetics and the sample is under flow beforehand in the first instance it would be open - I shall seek more clarity on this from Thomas
* Can we record all the timestamps for when TTL_x_ goes high/low?

**Comment by callumforrester** [comment]
@abbiemery can this be closed now or is there anything left to do?
